Release v0.4.4 (#253)
js-temporal/temporal-polyfill: CHANGELOG.md

# 0.4.4

Breaking changes:

> - `Temporal.ZonedDateTime.prototype.toLocaleString` now works without `Intl.DateTimeFormat`, and `Temporal.ZonedDateTime` objects are no longer supported as parameters to `Intl.DateTimeFormat` formatting methods. ([12071fb0], see also [Upstream PR](https://github.com/tc39/proposal-temporal/pull/2479), [Upstream PR 2](https://github.com/tc39/proposal-temporal/pull/2522))

Reviewer: What does "now works without `Intl.DateTimeFormat`" mean?

  - To format using `Temporal.ZonedDateTime`'s time zone, use `Temporal.ZonedDateTime.prototype.toLocaleString`. This method:
    1. Creates a new `Intl.DateTimeFormat` instance in the same time zone and calendar as the `Temporal.ZonedDateTime` instance
    1. Creates a `Temporal.Instant` from the `Temporal.ZonedDateTime` instance.
    1. Formats that `Temporal.Instant` instance using the created `Intl.DateTimeFormat` instance.
  - To format in a time zone that does not match the time zone of the `Temporal.ZonedDateTime` instance:
    1. Create a new `Temporal.ZonedDateTime` instance in the right time zone using `Temporal.ZonedDateTime.prototype.withTimeZone`.
       - To get the current system time zone, use `Temporal.Now.timeZoneId()`.

>          Note that caching this value can be very problematic because the system time zone can change during the lifetime of your program (e.g. when a user on a mobile device crosses a time zone boundary, users manipulate their system clock settings manually).

Reviewer: ```suggestion
         Be careful when caching the current system time zone (or an `Intl.DateTimeFormat` instance using that time zone), because the system time zone can change during the lifetime of your program (e.g. when a user on a mobile device crosses a time zone boundary or when a user manually edits time zone settings).
```

    1. Follow the steps above.
  - Generally, try to always include the `timeZone` property when creating an `Intl.DateTimeFormat` instance to avoid an ambiguity about which time zone the formatted results will be in.

>   - `Intl.DateTimeFormat instances` can be expensive to create. For performance-sensitive code where the same calendar and time zone are used repeatedly for formatting we recommend creating and reusing an `Intl.DateTimeFormat` instance with the desired `timeZone` and `calendar` options, and formatting using the value of `Temporal.ZonedDateTime.prototype.epochMilliseconds`.

Reviewer: ```suggestion
  - `Intl.DateTimeFormat` instances can be expensive to create. For performance-sensitive code where the same calendar and time zone are used repeatedly for formatting, we recommend creating and reusing an `Intl.DateTimeFormat` instance with the desired `timeZone` and `calendar` options, and then formatting using the value of `Temporal.ZonedDateTime.prototype.epochMilliseconds`.
```

    `Intl.DateTimeFormat` instances may also require a lot of RAM, so indefinitely caching large numbers of them is not recommended for memory-constrained environments.

> - The `.calendar` getter on Temporal types has been replaced with a `.calendarId` getter that returns the string calendar identifier, and a `.getCalendar()` method that returns an object that implements `Temporal.CalendarProtocol`.

Reviewer: ```suggestion
- The `.calendar` property on Temporal types has been replaced with a `.calendarId` property that returns the string calendar identifier and a `.getCalendar()` method that returns an object that implements `Temporal.CalendarProtocol`.
```

  - `getCalendar()` will always return a new object if the containing instance was constructed using a built-in calendar identifier string. However, if the containing instance was constructed using a calendar object (either a built-in calendar or custom calendar), then the same object is returned. ([cc701b45], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))

>   - Internally, Temporal will compare calendars only using their `.id` property. ([3916e547], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))

Reviewer: ```suggestion
  - Temporal will compare calendars by testing for string equality between calendar identifiers.
    When a calendar is an object, its `id` property will be used for comparison. ([3916e547], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))
```

  - The `.calendar` property has been entirely removed from `Temporal.PlainTime`. ([0727d86e], see also [Issue 1](https://github.com/tc39/proposal-temporal/issues/1808), [Issue 2](https://github.com/tc39/proposal-temporal/issues/1588))

> - The `.timeZone` getter on `Temporal.ZonedDateTime` has been replaced with a `.timeZoneId` getter that returns the string time zone identifier and a `.getTimeZone()` method that returns an object that implements `Temporal.TimeZoneProtocol`. ([d3263f0], see also Upstream issue)

Reviewer: ```suggestion
- The `.timeZone` property on `Temporal.ZonedDateTime` has been replaced with a `.timeZoneId` property that returns the string time zone identifier and a `.getTimeZone()` method that returns an object that implements `Temporal.TimeZoneProtocol`. ([d3263f0], see also Upstream issue)
```

  - `getTimeZone()` will always return a new object if the `Temporal.ZonedDateTime` instance was constructed using a built-in time zone identifier string.
    However, if the instance was constructed using a time zone object (either a built-in time zone or custom time zone), then the same object is returned.
  - `Termporal.Now.timeZone()` is replaced by `Temporal.Now.timeZoneId()`, which returns a time zone identifier string, not a `Temporal.TimeZone` instance. ([600f0cc7], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))

>   - Temporal compares `Temporal.TimeZones` only using their `.id` values, not the entire `ToString` representation. ([d6cb2862], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))

Reviewer: ```suggestion
  - Temporal will compare time zones by testing for string equality between calendar identifiers.
    When a calendar is an object, its `id` property will be used for comparison.  ([d6cb2862], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))
```

- Remove support for nested time zone and calendar property bags. ([9b7f61ae], [0c38267b], see also [Discussion](https://github.com/tc39/proposal-temporal/issues/2104#issuecomment-140))
- Require the remainder returned by the `NanosecondsToDays` internal operation to be less than the length of one calendar day.
  This could happen due to shenanigans in a custom time zone's `getOffsetNanosecondsFor` or `getPossibleInstantsFor` methods, or a custom calendar's `dateAdd` method.
  ([44b00a38], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/2357)
- Set an upper limit of 1e9 on rounding increments. ([5d78c815])
- Require `fields` and `mergeFields` methods for `Temporal.CalendarProtocol` (custom calendar) objects. ([5ad63274])

Bug fixes:

- Read field identifiers from a calendar's `fields()` method during `dateFromFields()`, `yearMonthFromFields()`, and `monthDayFromFields()`.
  Previously, these methods only referenced hardcoded `'era'` and `'eraYear'` field identifiers. ([375a4ad9])
- Avoid precision loss in AddDuration operations. ([c0f7349a])

> - Fix an infinite-loop bug and a RangeError during non-ISO calendar mathematics. ([d94c1cd9], see also [proposal-temporal PR](https://github.com/tc39/proposal-temporal/pull/2539), [Issue 1](https://github.com/tc39/proposal-temporal/issues/2383), [Issue 2](https://github.com/tc39/proposal-temporal/issues/2537))

Reviewer: ```suggestion
- Fix an infinite-loop bug and a RangeError during non-ISO calendar calculations. ([d94c1cd9], see also [proposal-temporal PR](https://github.com/tc39/proposal-temporal/pull/2539), [Issue 1](https://github.com/tc39/proposal-temporal/issues/2383), [Issue 2](https://github.com/tc39/proposal-temporal/issues/2537))
```

- Always start at the end of 1972 when computing a `Temporal.PlainMonthDay` from fields, preventing the reference year from accidentally being in 1971. ([ef4a0c4b])
- Apply the overflow behaviour to year/month/day values in `monthDayFromFields`. ([7ebd0f96])
- Preserve the day of month when constraining a nonexistent leap month, instead of defaulting to the end of the closest corresponding regular month. ([996f8fa1])
- Allow month codes `'M01L'` and `'M12L'` in the Chinese calendar. ([696f2c7e])
- Avoid overflows in `GetNamedTimeZoneOffsetNanoseconds`. ([c42570b8])
- Fix calendar validation in various ToTemporal\_\_\_ operations. ([e3913974], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/2546))
- Don't call `GetMethod` on a string calendar. ([fe698d8d], see also [Upstream issue](https://github.com/tc39/proposal-temporal/pull/2547))
- Avoid rounding errors in `BalanceDurationRelative` and `UnbalanceDurationRelative`. ([a907acf0])
- Check for negative day length in `Temporal.ZonedDateTime.prototype.round`. ([0d2d60ec], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/2261))

> - Reject Z designators when parsing `Temporal.PlainTime` strings.

Reviewer: I would move this into the Breaking Changes section.

- Add validation for the return values from calendar operations. ([d88cfa4d])
- Validate required methods of `Temporal.TimeZoneProtocol` and `Temporal.CalendarProtocol`. ([84563cea], [755c7620], see also [Discussion](https://github.com/tc39/proposal-temporal/issues/2104#issuecomment-140))
- Throw earlier when users might mix up positional `Temporal.TimeZone` and `Temporal.Calendar` arguments with each other, to prevent bugs like `new Temporal.ZonedDateTime(0n, cal, tz)` where the switched calendar and time zone arguments would cause exceptions to be thrown later. ([7922f1f9])

Non-breaking changes:

- Support new Internet Extended Date-Time (IXDTF) Annotations
  - See the [Temporal + IXDTF tracking issue](https://github.com/tc39/proposal-temporal/issues/1450).
  - Align ISO 8601 grammar with annotations from IXDTF specification.
    Calendar and time zone annotations are now allowed to contain a "critical" flag (`'!'`) prefix.
    Critical flags have no effect when parsing input strings because Temporal already treats unknown or inconsistent inputs as errors by default. ([e8b2e71c])

>     - There can be multiple annotations in a string, where unknown type annotations are ignored unless marked as critical.

Reviewer: ```suggestion
    - There can be multiple types of annotations in an IXDTF string.
      Temporal only recognizes time zone and calendar annotations.
      Unrecognized non-critical annotations will be ignored.
      Unrecognized critical annotations will cause the parsing method to throw an exception.
```

>   - Allow `toString()` methods for `Temporal.PlainDate`, `Temporal.PlainDateTime`, `Temporal.PlainYearMonth`, `Temporal.PlainMonthDay`, and `Temporal.ZonedDateTime` to emit critical IXDTF annotations using the "critical" option.

Reviewer: ```suggestion
  - Allow `toString()` methods for `Temporal.PlainDate`, `Temporal.PlainDateTime`, `Temporal.PlainYearMonth`, `Temporal.PlainMonthDay`, and `Temporal.ZonedDateTime` to emit critical IXDTF annotations using the `'critical'` option.
```

    Use this option with care, because the platform that you're communicating with may not understand this syntax.

>     `calendarName: 'critical'` behaves like `calendarName: 'always'` and `timeZoneName: 'critical'` behaves like `timeZoneName: 'always'`, but they also output a `'!'` flag in the corresponding annotation. ([50a64f16])

Reviewer: ```suggestion
    `calendarName: 'critical'` behaves like `calendarName: 'always'` and `timeZoneName: 'critical'` behaves like `timeZoneName: 'always'`, but they also output a `'!'` prefix in the corresponding annotation. ([50a64f16])
```

    Critical flags are never used by Temporal, but could be consumed by other programs.
  - Ignore calendar annotations when parsing `Temporal.Instant` strings. ([b86b87f0])

>   - Allow IXDTF annotations after YYYY-MM and MM-DD strings. UTC offsets in these annotations are forbidden and will cause an error to be thrown ([acd6464f])

Reviewer: ```suggestion
  - Allow calendar and/or time zone annotations after strings without a time part: YYYY-MM-DD, YYYY-MM, and MM-DD. ([acd6464f])
  - Disallow UTC offsets in YYYY-MM and MM-DD strings because they could cause ambiguity between an offset and the day of a YYYY-MM-DD string. ([acd6464f])
```
I think it's clearer if we split them.

  - Reject ambiguous time strings even with calendar annotations. ([af875275])
- Implement the full set of rounding modes.
  New modes include `'expand'`, `'halfCeil'`, `'halfFloor'`, `'halfTrunc'`, and `'halfEven'`.
  ([eb5404d1])
- Treat calendar names as case-insensitive. ([9e730d68])

> - Implement `yearOfWeek` methods that complement existing `weekOfYear` methods.

Reviewer: I would move this up to the top of the list, as I think we've gotten a lot of requests for this.

# 0.4.4

Breaking changes:

- ZonedDateTime.toLocaleString now works without Intl.DateTimeFormat, and Intl.DateTimeFormat objects are no longer supported as parameters to Intl.DateTimeFormat methods. ([12071fb0], see also [Upstream PR](https://github.com/tc39/proposal-temporal/pull/2479), [Upstream PR 2](https://github.com/tc39/proposal-temporal/pull/2522))
  - To format using the ZonedDateTime's timezone use `ZonedDateTime.prototype.toLocaleString`, which creates a new Intl.DateTimeFormat instance in the same timezone as the ZonedDateTime and formats a Temporal.Instant created from the ZonedDateTime. For performance-sensitive code, we recommend that the same thing be done manually to re-use the created DateTimeFormat instance as they can be very expensive to create.

>   - To format using the Host System's timezone, convert the ZonedDateTime into an Instant and pass the Instant to a Intl.DateTimeFormat without specifying a timezone in the DateTimeFormat constructor.

Author: PTAL at the latest version and let me know what you think.

- .calendar getters on Temporal types have been replaced with .calendarId getters that always return the string calendar identifier value, and a .getCalendar() method that always returns a calendar object.
  - getCalendar() will always return a new object if the calendar is an internal calendar. ([cc701b45], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))
  - Internally, Temporal will compare calendars only using their .id property. ([3916e547], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/1808))

>   - The calendar property has been entirely removed from the PlainTime type. ([0727d86e], see also [Issue 1](https://github.com/tc39/proposal-temporal/issues/1808), [Issue 2](https://github.com/tc39/proposal-temporal/issues/1588))

Author: This is making the commit ID both the text and the Markdown link specificer. When I render this markdown using the vscode markdown renderer they just appear as the commit ID, but linkified (to the commit on github).

>   - The calendar property has been entirely removed from the PlainTime type. ([0727d86e], see also [Issue 1](https://github.com/tc39/proposal-temporal/issues/1808), [Issue 2](https://github.com/tc39/proposal-temporal/issues/1588))

Author: (which works because the link specifier is then matched to the corresponding markdown link reference later at the bottom of the changelog).

- Use BigInt math in RoundDuration to avoid problems when the values are larger than MAX_SAFE_INTEGER. ([955323f8])
- Always start at the end of 1972 when computing a MonthDay from fields, preventing the reference year from accidentally being in 1971. ([ef4a0c4b])
- Apply the overflow behaviour to year/month/day values in monthDayFromFields. ([7ebd0f96])
- Preserve the day of month when constraining a nonexistent leap month, instead of defaulting to the end of the closest corresponding regular month. ([996f8fa1])
- Allow months M01L and M12L in Chinese calendar. ([696f2c7e])
- Avoid overflows in GetNamedTimeZoneOffsetNanoseconds. ([c42570b8])
- Fix calendar validation in various ToTemporal\_\_\_ operations. ([e3913974], see also [Upstream issue](https://github.com/tc39/proposal-temporal/issues/2546))
- Don't call GetMethod on a string calendar. ([fe698d8d], see also [Upstream issue](https://github.com/tc39/proposal-temporal/pull/2547))
- Avoid rounding errors in (Un)balanceDurationRelative. ([a907acf0])
- Check for negative day length in ZonedDateTime.round. ([0d2d60ec], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/2261))

> - Reject Z designators when parsing PlainTime strings. ([a7a50eac])

Reviewer: ```suggestion
- Reject Z designators when parsing `Temporal.PlainTime` strings.
  This change avoids a common bug where a UTC timestamp is accidentally interpreted as if it were a local time in a real time zone.
  If you do want to parse the time portion of a UTC timestamp string, use: `Temporal.Instant.from(s).toZonedDateTimeISO('UTC').toPlainTime()`. ([a7a50eac])
```

> - Reject 0-value components when parsing Duration strings, and avoid rounding errors when nanoseconds components are present. ([58b5601a])

Reviewer: ```suggestion
- Reject 0-value components when parsing `Temporal.Duration` strings, and avoid rounding errors when nanoseconds components are present. ([58b5601a])
```

> - Reject 0-value components when parsing Duration strings, and avoid rounding errors when nanoseconds components are present. ([58b5601a])

Reviewer: ```suggestion
- Reject 0-value components when parsing `Temporal.Duration` strings, and avoid rounding errors when nanoseconds components are present. ([58b5601a])
```

> - Reject relativeTo string arguments that are neither valid PlainDate nor ZonedDateTime strings (such as "2022-08-18T17:01Z"). ([4db15c41])

Reviewer: ```suggestion
- Reject `relativeTo` string options that are neither valid `Temporal.PlainDate` nor `Temporal.ZonedDateTime` strings, such as `'2022-08-18T17:01Z'`. ([4db15c41])
```

> - Reject relativeTo string arguments that are neither valid PlainDate nor ZonedDateTime strings (such as "2022-08-18T17:01Z"). ([4db15c41])

Reviewer: ```suggestion
- Reject `relativeTo` string options that are neither valid `Temporal.PlainDate` nor `Temporal.ZonedDateTime` strings, such as `'2022-08-18T17:01Z'`. ([4db15c41])
```

- Add validation for the return values from calendar operations. ([d88cfa4d])

> - Validate required methods of TimeZone and Calendar protocols. ([84563cea], [755c7620], see also [Discussion](https://github.com/tc39/proposal-temporal/issues/2104#issuecomment-140))

Reviewer: ```suggestion
- Validate required methods of `Temporal.TimeZoneProtocol` and `Temporal.CalendarProtocol`. ([84563cea], [755c7620], see also [Discussion](https://github.com/tc39/proposal-temporal/issues/2104#issuecomment-140))
```
